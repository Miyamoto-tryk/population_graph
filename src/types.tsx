//年と人口。APIレスポンスの、各年の人口を表す型
export type PeoplAndYear = {
  year: number;
  value: number;
};
//人口データ
export type PoplData = {
  year: number;
  value: number;
  PrefName: PrefecName;
  //DisplayData型へ変換しやすいようにPrefName型も保持している
};
//選択中の都道府県の、すべての人口データを、
//GraphData[]として保持する
export type GraphData = {
  label: DisplayLabel;
  prefCode: number;
  data: PoplData[];
};

//都道府県情報(都道府県名、都道府県コード)
export type Prefecture = {
  prefCode: number;
  prefName: PrefecName; //"北海道"or"青森県"or...
};
//都道府県チェックボックスの情報
export type PrefectureInfo = {
  prefecture: Prefecture;
  isSelected: boolean;
};

//APIレスポンスに含まれる各種人口データの型
export type ComposeDatas = {
  data: PeoplAndYear[]; //[{year , value}, ...]
  label: DisplayLabel; //"総人口"or"年少人口"or...
};
//都道府県別APIレスポンスの返り値型
export type ComposResponse = {
  boundaryYear: number; //2020
  data: ComposeDatas[]; //[0]:総人口,[1]:年少人口,[2]:生産年齢人口,[[3]:老年人口
};

//動的なプロパティ(都道府県名)を持つ。
//recharts(グラフ描画ライブラリ)に渡すためのデータ型
export type DisplayData = {
  [prop in PrefecName]?: number;
} & {
  year: number;
};
//プロパティと人口データの追加
export const addPoplData = (
  displayData: DisplayData[], //元のデータ
  poplData: PoplData[] //追加するデータ
): DisplayData[] => {
  const newData = displayData.map((elm: DisplayData, index: number) => {
    const prefecName = poplData[index].PrefName;
    const poplNum = poplData[index].value;
    return {
      ...elm,
      [prefecName]: poplNum,
    };
  });
  return newData;
};

export type DisplayLabel = "総人口" | "年少人口" | "生産年齢人口" | "老年人口";
export type PrefecName =
  | "北海道"
  | "青森県"
  | "岩手県"
  | "宮城県"
  | "秋田県"
  | "山形県"
  | "福島県"
  | "茨城県"
  | "栃木県"
  | "群馬県"
  | "埼玉県"
  | "千葉県"
  | "東京都"
  | "神奈川県"
  | "新潟県"
  | "富山県"
  | "石川県"
  | "福井県"
  | "山梨県"
  | "長野県"
  | "岐阜県"
  | "静岡県"
  | "愛知県"
  | "三重県"
  | "滋賀県"
  | "京都府"
  | "大阪府"
  | "兵庫県"
  | "奈良県"
  | "和歌山県"
  | "鳥取県"
  | "島根県"
  | "岡山県"
  | "広島県"
  | "山口県"
  | "徳島県"
  | "香川県"
  | "愛媛県"
  | "高知県"
  | "福岡県"
  | "佐賀県"
  | "長崎県"
  | "熊本県"
  | "大分県"
  | "宮崎県"
  | "鹿児島県"
  | "沖縄県";
