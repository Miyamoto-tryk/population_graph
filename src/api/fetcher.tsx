import { Prefecture, ComposResponse, ComposeDatas } from "../types";

const api_key = process.env.NEXT_PUBLIC_RESAS_API_KEY;
const headers = new Headers();
if (api_key) {
  headers.append("X-API-KEY", api_key);
}

//都道府県一覧 APIから取得
export const fetchPrefec = async (): Promise<Prefecture[]> => {
  const res = await fetch(
    "https://opendata.resas-portal.go.jp/api/v1/prefectures",
    { method: "GET", headers: headers }
  );
  //TODO: 型ガード関数の実装
  const prefectures = await res.json();
  console.log(prefectures);
  //console.log(prefectures.result);
  return prefectures.result;
};

//人口構成をAPIから取得
export const fetchCompos = async (
  prefCode: number
): Promise<ComposeDatas[]> => {
  if (prefCode > 47 || prefCode < 1) {
    throw new Error("prefecture code が不正です");
  }
  const res = await fetch(
    `https://opendata.resas-portal.go.jp/api/v1/population/composition/perYear?cityCode=-&prefCode=${prefCode}`,
    { method: "GET", headers: headers }
  );
  const ans = await res.json();
  console.log(ans);
  const compos: ComposeDatas[] = Object.values(ans.result.data);
  //console.log(compos);
  return compos;
};
