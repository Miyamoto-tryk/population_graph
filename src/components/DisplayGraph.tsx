import {
  CartesianGrid,
  Legend,
  Line,
  LineChart,
  Tooltip,
  XAxis,
  YAxis,
} from "recharts";

import {
  DisplayData,
  DisplayLabel,
  PrefecName,
  PrefectureInfo,
} from "../types";
import { Dispatch, SetStateAction, useEffect, useState } from "react";
import { GraphData } from "../types";
import { ResponsiveContainer } from "recharts";
import { extractLabelData } from "../data_controller/extractLabelData";
import {
  addPrefecColor,
  colorPrefecMap,
} from "../data_controller/colorPrefecMap";
import styles from "@/app/page.module.css";
type Props = {
  displayLabel: DisplayLabel;
  graphData: GraphData[];
  displayData: DisplayData[];
  selectedPrefec: PrefecName[];
  setDisplayData: Dispatch<SetStateAction<DisplayData[]>>;
};

const DisplayGraph = ({
  displayLabel,
  graphData,
  displayData,
  selectedPrefec,
  setDisplayData,
}: Props) => {
  //表示データの抽出・更新
  useEffect(() => {
    console.log(typeof graphData);
    const newDisplayData = extractLabelData({ graphData, displayLabel });
    setDisplayData(newDisplayData);
  }, [graphData, displayLabel]);

  if (graphData.length === 0) {
    return (
      <div
        style={{
          display: "flex",
          height: 400,
          alignItems: "center",
          justifyContent: "center",
        }}
      >
        <h3>Please select prefectures</h3>
      </div>
    );
  }

  //表示データが揃っていない場合
  if (Object.keys(displayData[0]).length <= 1 || selectedPrefec.length === 0) {
    return (
      <div
        style={{
          display: "flex",
          height: 400,
          alignItems: "center",
          justifyContent: "center",
        }}
      >
        <h3>データ取得中</h3>
      </div>
    );
  }
  console.log(selectedPrefec);
  //グラフの描画
  return (
    <>
      <div style={{ display: "flex", justifyContent: "center" }}>
        <ResponsiveContainer width="90%" height={400}>
          <LineChart data={displayData}>
            <CartesianGrid strokeDasharray="3 3" />
            <XAxis dataKey="year" />
            <YAxis />
            {selectedPrefec.map((prefecName: PrefecName) => {
              //都道府県の色が記録されていない場合、追加
              if (!colorPrefecMap.has(prefecName)) addPrefecColor(prefecName);
              return (
                <Line
                  key={prefecName}
                  type="monotone"
                  dataKey={prefecName}
                  stroke={colorPrefecMap.get(prefecName)}
                />
              );
            })}
            <Legend />
            <Tooltip />
          </LineChart>
        </ResponsiveContainer>
      </div>
    </>
  );
};

export default DisplayGraph;
