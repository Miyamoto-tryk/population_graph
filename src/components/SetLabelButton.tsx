import { DisplayData, DisplayLabel } from "@/types";
import { Dispatch, SetStateAction } from "react";
import styles from "@/app/page.module.css";

type Props = {
  setDisplayLabel: Dispatch<SetStateAction<DisplayLabel>>;
};

//ラベルの種類
const displayLabels: DisplayLabel[] = [
  "総人口",
  "年少人口",
  "生産年齢人口",
  "老年人口",
];

//ボタンを表示
const SetLabelButton = ({ setDisplayLabel }: Props) => {
  const handleClick = (label: DisplayLabel) => {
    //ボタンが押されたらラベルを設定
    setDisplayLabel(label);
  };

  return (
    <>
      <div
        style={{
          display: "flex",
          flexDirection: "row",
          justifyContent: "center",
          gap: "10px",
        }}
      >
        {/* ボタンを並べる */}
        {displayLabels.map((label: DisplayLabel) => (
          <div key={label} dir="column">
            <button onClick={() => handleClick(label)}>{label}</button>{" "}
          </div>
        ))}
      </div>
    </>
  );
};

export default SetLabelButton;
