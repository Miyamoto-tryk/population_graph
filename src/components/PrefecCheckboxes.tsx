import React, { useState } from "react";
import { fetchCompos, fetchPrefec } from "../api/fetcher";
import { getGraphData } from "../data_controller/getGraphData";
import { getSelectedPrefec } from "../data_controller/getSelectedPrefec";
import { PrefectureInfo, Prefecture, GraphData, PrefecName } from "../types";
import { Dispatch, SetStateAction, FC, useEffect } from "react";

import styles from "../app/page.module.css";

type Props = {
  graphData: GraphData[];
  selectedPrefec: PrefecName[];
  setGraphData: Dispatch<SetStateAction<GraphData[]>>;
  setSelectedPrefec: Dispatch<SetStateAction<PrefecName[]>>;
};

//都道府県一覧を表示
export const PrefecCheckboxes: FC<Props> = ({
  graphData,
  selectedPrefec,
  setGraphData,
  setSelectedPrefec,
}: Props) => {
  //都道府県の選択情報
  const [prefecInfo, setPrefecInfo] = useState<PrefectureInfo[]>([]);
  //都道府県一覧を取得し、選択情報を初期化
  useEffect(() => {
    fetchPrefec().then((prefecData) => {
      const newPrefectures: PrefectureInfo[] = prefecData.map(
        (elm: Prefecture) => {
          //未選択で初期化
          return { prefecture: elm, isSelected: false };
        }
      );
      setPrefecInfo(newPrefectures);
    });
  }, []);

  const onCheckboxClicked = async (prefecCode: number) => {
    //都道府県のチェックボックスの情報を更新
    const updatedPrefecInfo: PrefectureInfo[] = prefecInfo.map(
      (elm: PrefectureInfo) => {
        if (elm.prefecture.prefCode == prefecCode) {
          return { prefecture: elm.prefecture, isSelected: !elm.isSelected };
        }

        return elm;
      }
    );

    //選択中の都道府県名を更新
    const updatedSelectedPrefec = getSelectedPrefec({
      prefecCode,
      prefecInfo,
      selectedPrefec,
    });

    //表示される可能性のあるグラフデータを更新
    const updatedGraphData = await getGraphData({
      prefecCode,
      prefecInfo,
      graphData,
    });

    setPrefecInfo(updatedPrefecInfo);
    setSelectedPrefec(updatedSelectedPrefec);
    setGraphData(updatedGraphData);
  };

  //都道府県情報が空である場合
  if (prefecInfo.length === 0) {
    return (
      <div>
        <h2>都道府県を取得中...</h2>
      </div>
    );
  }
  return (
    <>
      <div>
        <h2>都道府県一覧</h2>
        <div
          style={{
            display: "grid",
            gridTemplateColumns: "repeat(auto-fill, 100px)",
          }}
        >
          {prefecInfo.map((prefectureInfo: PrefectureInfo) => (
            //都道府県一覧とチェックボックスを生成
            <label key={prefectureInfo.prefecture.prefName}>
              <input
                type="checkbox"
                id={prefectureInfo.prefecture.prefName}
                checked={prefectureInfo.isSelected}
                onChange={() =>
                  onCheckboxClicked(prefectureInfo.prefecture.prefCode)
                }
              />
              <div style={{ paddingLeft: "4px", display: "inline" }}>
                {prefectureInfo.prefecture.prefName}
              </div>
            </label>
          ))}
        </div>
      </div>
    </>
  );
};
