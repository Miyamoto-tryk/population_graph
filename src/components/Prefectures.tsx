import { fetchPrefec } from "@/api/fetcher";
import { newGraphData } from "@/data_controller/newGraphData";
import { newSelectedPrefec } from "@/data_controller/newSelectedPrefec";
import { PrefectureInfo, Prefecture, GraphData, PrefecName } from "@/types";
import { Dispatch, SetStateAction, FC, useEffect } from "react";

type Props = {
  prefec_info: PrefectureInfo[];
  graph_data: GraphData[];
  selected_prefec: PrefecName[];
  setPrefec_info: Dispatch<SetStateAction<PrefectureInfo[]>>;
  setGraph_data: Dispatch<SetStateAction<GraphData[]>>;
  setSelected_prefec: Dispatch<SetStateAction<PrefecName[]>>;
};

//都道府県一覧を表示
const Prefectures: FC<Props> = ({
  prefec_info,
  graph_data,
  selected_prefec,
  setPrefec_info,
  setGraph_data,
  setSelected_prefec,
}: Props) => {
  //都道府県一覧を取得し、選択情報を初期化
  useEffect(() => {
    fetchPrefec().then((prefec_data) => {
      const newprefectures: PrefectureInfo[] = prefec_data.map(
        (elm: Prefecture) => {
          const selected: boolean = false; //未選択で初期化
          return { prefecture: elm, isSelected: selected };
        }
      );
      setPrefec_info(newprefectures);
    });
  }, []);

  const HandleChange = async (prefec_code: number) => {
    //都道府県の選択情報を更新
    const new_prefec_info: PrefectureInfo[] = prefec_info.map(
      (elm: PrefectureInfo) => {
        if (elm.prefecture.prefCode == prefec_code) {
          return { prefecture: elm.prefecture, isSelected: !elm.isSelected };
        } else {
          return elm;
        }
      }
    );
    setPrefec_info(new_prefec_info);
    const new_selected_prefec = newSelectedPrefec({
      prefec_code,
      prefec_info,
      selected_prefec,
    });
    setSelected_prefec(new_selected_prefec);

    const new_graph_data = await newGraphData({
      prefec_code,
      prefec_info,
      graph_data,
    });
    setGraph_data(new_graph_data);
  };

  if (!prefec_info) {
    return (
      <div>
        <h2>都道府県を取得中...</h2>
      </div>
    );
  }
  return (
    <>
      <div>
        <h2>都道府県一覧</h2>
        <div>
          {prefec_info.map((prefecInfo: PrefectureInfo) => (
            //都道府県一覧とチェックボックスを生成
            <div key={prefecInfo.prefecture.prefCode}>
              <input
                type="checkbox"
                checked={prefecInfo.isSelected}
                onChange={() => HandleChange(prefecInfo.prefecture.prefCode)}
              />
              {prefecInfo.prefecture.prefName}
            </div>
          ))}
        </div>
      </div>
    </>
  );
};
export default Prefectures;
