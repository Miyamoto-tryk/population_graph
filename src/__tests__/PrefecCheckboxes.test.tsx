import React, { act } from "react";
import {
  render,
  screen,
  waitFor,
  fireEvent,
  getByLabelText,
} from "@testing-library/react";
import { PrefecCheckboxes } from "../components/PrefecCheckboxes";
import { fetchPrefec, fetchCompos } from "../api/fetcher";
import TestRender from "react-test-renderer";
import "@testing-library/jest-dom";

import { GraphData, PrefecName, PrefectureInfo } from "../types";

// fetchPrefecをモック
jest.mock("../api/fetcher");

// モックデータ
const mockGraphData: GraphData[] = [];
const mockSelectedPrefec: PrefecName[] = [];
const mockSetGraphData = jest.fn();
const mockSetSelectedPrefec = jest.fn();

const mockPrefecInfo: PrefectureInfo[] = [
  { prefecture: { prefCode: 1, prefName: "北海道" }, isSelected: false },
  { prefecture: { prefCode: 2, prefName: "青森県" }, isSelected: false },
];

describe("Prefectures Component", () => {
  beforeEach(() => {
    // モックをリセット
    jest.clearAllMocks();
  });

  test("PrefecCheckboxes matches snapshot", () => {
    const component = TestRender.create(
      <PrefecCheckboxes
        graphData={mockGraphData}
        selectedPrefec={mockSelectedPrefec}
        setGraphData={mockSetGraphData}
        setSelectedPrefec={mockSetSelectedPrefec}
      />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders loading message initially", async () => {
    (fetchPrefec as jest.Mock).mockResolvedValue([]);

    waitFor(() => {
      render(
        <PrefecCheckboxes
          graphData={[]}
          selectedPrefec={[]}
          setGraphData={mockSetGraphData}
          setSelectedPrefec={mockSetSelectedPrefec}
        />
      );
    });

    expect(screen.getByText("都道府県を取得中...")).toBeInTheDocument();
  });

  it("renders prefecture list after fetch", async () => {
    (fetchPrefec as jest.Mock).mockResolvedValue([
      { prefCode: 1, prefName: "北海道" },
      { prefCode: 2, prefName: "青森県" },
    ]);
    const mockSetPrefec = jest.fn();
    jest
      .spyOn(React, "useState")
      .mockImplementationOnce(() => [[], mockSetPrefec]);

    render(
      <PrefecCheckboxes
        graphData={[]}
        selectedPrefec={[]}
        setGraphData={mockSetGraphData}
        setSelectedPrefec={mockSetSelectedPrefec}
      />
    );

    await waitFor(() => expect(mockSetPrefec).toHaveBeenCalledTimes(1));
    expect(mockSetPrefec).toHaveBeenCalledWith(mockPrefecInfo);
  });
});
