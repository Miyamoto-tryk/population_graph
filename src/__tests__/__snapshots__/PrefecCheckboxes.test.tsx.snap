// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prefectures Component PrefecCheckboxes matches snapshot 1`] = `
<div>
  <h2>
    都道府県を取得中...
  </h2>
</div>
`;
